[css-animations-2] Specify the animation-trigger property

ydaniv opened the pull request:
Adding initial draft for Animation Triggers and the `animation-trigger` property as resolved in https://github.com/w3c/csswg-drafts/issues/8942#issuecomment-1854342340.

I cleared all the issues and fleshed out the gist of it.

cc @flackr @birtles @dbaron

flackr commented on css-animations-2/Overview.bs:
```diff
@@ -847,6 +873,271 @@ expressed in milliseconds, it must be divided by 1,000 to produce a value in
+An <dfn export>animation trigger</dfn> is used to control the playback
+of its associated [=animation=] for time-driven animations.
+[=Animation triggers=] follow the same [[web-animations-1#hierarchical|hierarchy]]
+as animations and are associated with a [=timeline=] and an [=animation trigger effect|effect=].
```
This makes it sound like animation triggers are separate things from animations, but I suspect we should say that as a part of [updating the current time of animations](https://drafts.csswg.org/web-animations-1/#update-animations-and-send-events) we evaluate whether a trigger condition has been reached and apply a stateful change to the animation.

ydaniv replied:
Sounds good. I'll work on that

flackr commented on css-animations-2/Overview.bs:
```diff
@@ -847,6 +873,271 @@ expressed in milliseconds, it must be divided by 1,000 to produce a value in
+[=Animation triggers=] follow the same [[web-animations-1#hierarchical|hierarchy]]
+as animations and are associated with a [=timeline=] and an [=animation trigger effect|effect=].
+
+Issue: Should there be any effect of triggers on scroll-driven animations?
```
I suspect triggers on scroll-driven animations should "just work". E.g. they could play / stop the animation when the trigger condition occurs / finishes.

ydaniv replied:
How does each `type` affects a scroll-driven animation?
Furthermore, what could be the actual use-case for this? Perhaps one that I can think of now is if we have a `hover-timeline` in the future and we want to enable/disable it on different scroll positions. Is that what you mean?

flackr commented on css-animations-2/Overview.bs:
```diff
@@ -847,6 +873,271 @@ expressed in milliseconds, it must be divided by 1,000 to produce a value in
+  ::
+
+    <dl class=switch>
+      : If |effect| is inside its [=active interval=],
```
I think defining a trigger in terms of an internal animation effect's time might be overkill. E.g. I think we could define this simply in terms of the animation timeline's time with respect to the specified range

ydaniv replied:
I struggled with that quite a bit. Eventually it worked out once I defined the same hierarchy as animations', and defined a special state for the `effect`. I'll try simplifying this further.

flackr commented on css-animations-2/Overview.bs:
```diff
@@ -847,6 +873,271 @@ expressed in milliseconds, it must be divided by 1,000 to produce a value in
+
+Issue: Should there be any effect of triggers on scroll-driven animations?
+
+##  Animation Trigger Effects ## {#trigger-effects}
```
I think the internal control of how triggers work probably belongs in web-animations where we define things like how an updated time updates corresponding animations: https://drafts.csswg.org/web-animations-1/#update-animations-and-send-events

We should also be able to set up animation triggers from javascript.

ydaniv replied:
OK, sure. We didn't discuss the WAAPI part of triggers in the issue, should I try fleshing that too in this PR? Or you're just stating this as a reason for this change?

flackr reviewed (commented):
Thanks for getting this started! I think some of the low level details of how an animation trigger affects the animation probably belong in web-animations-1 with the css-animations spec setting properties by which you set up those triggers declaratively. You should also of course be able to set the same properties from the web-animations api.

ydaniv commented on css-animations-2/Overview.bs:
```diff
@@ -690,7 +693,30 @@ The 'animation' shorthand property syntax is as follows:
 
 <span class=prod><dfn>&lt;single-animation></dfn> = <<'animation-duration'>> || <<easing-function>> || <<'animation-delay'>> || <<single-animation-iteration-count>> || <<single-animation-direction>> || <<single-animation-fill-mode>> || <<single-animation-play-state>> || [ none | <<keyframes-name>> ] || <<single-animation-timeline>></span>
 
+## The 'animation-trigger' property ## {#animation-trigger}
```
I started with everything next to the short-hand, but later had to move more and properties close to the definition of Trigger below. I'll refactor it now anyway, to have the properties here, and the definitions in web-animaitons-1, as you said.

ydaniv commented:
@birtles thanks for reviewing!

The intention of the hierarchy was mostly to simplify the spec, by having triggers follow animations as in:
For animations: ` timeline -> animation -> effect`
For triggers: `timeline -> trigger -> effect`
And since effects are attached to attachment ranges, have trigger follow the same mechanism.

The idea was that they don't share anything, just follow the same structure, and the trigger simply affects the animation's playback according to its properties and state.

But, as @flackr mentioned, it's still too confusing and too complex, so I'll try and rewrite this without the effect part, and only with a trigger attached to a timeline and a range.

I've tried defining the constructs first in the initial draft. I'll move those into the Web Animations spec as suggested, and try to organize them together again in a simpler way.

birtles commented:
> The intention of the hierarchy was mostly to simplify the spec, by having triggers follow animations as in:
> For animations: ` timeline -> animation -> effect`
> For triggers: `timeline -> trigger -> effect`
> And since effects are attached to attachment ranges, have trigger follow the same mechanism.
> 
> The idea was that they don't share anything, just follow the same structure, and the trigger simply affects the animation's playback according to its properties and state.

Makes sense!

> But, as @flackr mentioned, it's still too confusing and too complex, so I'll try and rewrite this without the effect part, and only with a trigger attached to a timeline and a range.

Ok, I'll look forward to seeing what you come up with!

ydaniv commented:
@flackr @birtles  I've just pushed changes, I think I've covered all the comments, thanks!

birtles commented on web-animations-1/Overview.bs:
```diff
@@ -3537,6 +3540,141 @@ Calculating progress {#core-animation-effect-calculations}
+<dl dfn-for="animation trigger state" dfn-type=value>
+  <dt><dfn>idle</dfn>
+  <dd>
+    The [=animation effect=] associated with |trigger| remains in
+    its [=animation effect/before phase=] and stays at zero [=animation/start time=].
```
I still don't quite get the hierarchy here. This suggests an animation effect is associated with a trigger but previous we said a trigger is associated with an animation. Which is it?

ydaniv replied:
Right, the relation here should be indirect via the animation. The trigger is associated to the animation. The required behavior of the animation is to "hold its effect in a 'just before playing' state". I'll try to fix that.

ydaniv replied:
Done, I think it's better now.

ydaniv commented:
kind ping @birtles @flackr, ready for another review.

birtles commented:
> kind ping @birtles flackr, ready for another review.

@ydaniv Sorry for the delay--currently on parental leave. Just a few meta points:

1. I think this should be part of Web Animations level 2. Level 1 is basically done except for fixing existing issues (most notably the `commitStyles` changes we agreed to but I've yet to edit in).
2. I think we need to define the API parts for triggers?
3. We probably need to review all the descriptions about timelines and animations to make sure it accounts for triggers. (e.g. the sentence "Updating the [current time](https://drafts.csswg.org/web-animations-1/#animation-current-time) of any [animations](https://drafts.csswg.org/web-animations-1/#concept-animation) [associated with](https://drafts.csswg.org/web-animations-1/#associated-with-a-timeline) the timeline." should probably refer to triggers too?)

ydaniv commented:
Thanks @birtles!

> currently on parental leave

Congrats ^_^!

> 1. I think this should be part of Web Animations level 2.

Ok, I also wondered about that, but @flackr wrote Web Animations 1, and some things I needed to update were actually there, so not sure whether this was intentional or not.

> 2. I think we need to define the API parts for triggers?

You mean the DOM interfaces, as in [here](https://drafts.csswg.org/css-animations-1/#interface-dom)?

> 3. We probably need to review all the descriptions about timelines and animations to make sure it accounts for triggers. (e.g. the sentence "Updating the [current time](https://drafts.csswg.org/web-animations-1/#animation-current-time) of any [animations](https://drafts.csswg.org/web-animations-1/#concept-animation) [associated with](https://drafts.csswg.org/web-animations-1/#associated-with-a-timeline) the timeline." should probably refer to triggers too?)

I've added [this text](https://github.com/w3c/csswg-drafts/pull/10128/files#diff-c240da42474be67dd65a2e91af4774da3cd91e832e1a6401b9fbefd6efe18851R775) to that section:

> * Updating the [=animation trigger state=] of any [=animation trigger=]
              [=associated with a timeline|associated with=] the timeline.

Shouldn't this cover it? I think basically comes down to checking triggers state on timeline's current time update.

birtles commented:
> > currently on parental leave
> 
> Congrats ^_^!

Thanks!

> > 1. I think this should be part of Web Animations level 2.
> 
> Ok, I also wondered about that, but @flackr wrote Web Animations 1, and some things I needed to update were actually there, so not sure whether this was intentional or not.

Yeah, that would be good to know. My understanding is it should be level 2 but maybe @flackr has some ideas about that.

> > 2. I think we need to define the API parts for triggers?
> 
> You mean the DOM interfaces, as in [here](https://drafts.csswg.org/css-animations-1/#interface-dom)?

I was thinking about the Web Animations DOM interfaces. Just like we have the [`Animation` interface](https://drafts.csswg.org/web-animations-1/#the-animation-interface) I assume we'll have an `AnimationTrigger` interface?

That also makes me wonder how authors would expect to be able to inspect these from script. Is there some way to fetch the triggers in a document? Or the trigger driving a particular `Animation`?
 
> > 3. We probably need to review all the descriptions about timelines and animations to make sure it accounts for triggers. (e.g. the sentence "Updating the [current time](https://drafts.csswg.org/web-animations-1/#animation-current-time) of any [animations](https://drafts.csswg.org/web-animations-1/#concept-animation) [associated with](https://drafts.csswg.org/web-animations-1/#associated-with-a-timeline) the timeline." should probably refer to triggers too?)
> 
> I've added [this text](https://github.com/w3c/csswg-drafts/pull/10128/files#diff-c240da42474be67dd65a2e91af4774da3cd91e832e1a6401b9fbefd6efe18851R775) to that section:
> 
> > * Updating the [=animation trigger state=] of any [=animation trigger=]
> >   [=associated with a timeline|associated with=] the timeline.
> 
> Shouldn't this cover it? I think basically comes down to checking triggers state on timeline's current time update.

Yeah, I was thinking that we need to review the whole section to check for other references that need to be updated like the sentence I mentioned and other definitions like this one:

> An [animation effect](https://drafts.csswg.org/web-animations-1/#animation-effect), effect, is associated with a timeline, timeline, if effect is [associated with an animation](https://drafts.csswg.org/web-animations-1/#animation-effect-associated-animation) which, in turn, is associated with timeline.

Presumably that needs to be updated to include triggers too?

This is a pretty fundamental change to the whole timing hierarchy so we need to be careful about it. I wonder if it would be better to introduce an abstract concept that covers both animations and animation triggers like "timeline subscribers" or somesuch.

ydaniv commented:
> I was thinking about the Web Animations DOM interfaces. Just like we have the [Animation interface](https://drafts.csswg.org/web-animations-1/#the-animation-interface) I assume we'll have an AnimationTrigger interface?

I opened an issue with a rough proposal [here](https://github.com/w3c/csswg-drafts/issues/10229).
I could add that in already, as an initial proposal for API. I thought to wait with that for a later PR, since we only resolved on adding an initial draft and going with it back to the group for approval.

> That also makes me wonder how authors would expect to be able to inspect these from script. Is there some way to fetch the triggers in a document? Or the trigger driving a particular Animation?

I missed that part, I guess I could add that in. Regarding additional APIs, like getting all triggers from document, I guess we could always add these later on demand?

> Yeah, I was thinking that we need to review the whole section to check for other references that need to be updated like the sentence I mentioned and other definitions like this one:
> 
>> An [animation effect](https://drafts.csswg.org/web-animations-1/#animation-effect), effect, is associated with a timeline, timeline, if effect is [associated with an animation](https://drafts.csswg.org/web-animations-1/#animation-effect-associated-animation) which, in turn, is associated with timeline.
> 
> Presumably that needs to be updated to include triggers too?

I checked everything and didn't find anything else. I think since triggers only affect playback state and are associated with a timeline, it should be enough to check/update their state according to the timeline, since the trigger's spec describes what to do on state change.

>This is a pretty fundamental change to the whole timing hierarchy so we need to be careful about it. I wonder if it would be better to introduce an abstract concept that covers both animations and animation triggers like "timeline subscribers" or somesuch.

Agreed. I tried simplifying it on the last change.

flackr commented:
> > Ok, I also wondered about that, but @flackr wrote Web Animations 1, and some things I needed to update were actually there, so not sure whether this was intentional or not.

> Yeah, that would be good to know. My understanding is it should be level 2 but maybe @flackr has some ideas about that.

Sorry if I created some confusion here. I was mostly getting at that the concepts that need updating are in the web-animations-1 spec, but of course additions to the spec should go in web-animations-2.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2354,142 @@ Items sorted earlier are executed before those sorted later.
+<dl class=switch>
+  : If |trigger|’s [=local time=] is [=unresolved=],
+  ::
+    Then |state| is ''animation trigger state/idle'', and |did trigger| is set to false.
```
Wouldn't setting `did trigger` to false this reset `once` triggers anytime the timeline goes idle (e.g. anytime a ScrollTimeline's scroller becomes large enough that it doesn't scroll)? I think this is probably unexpected.

Note, it doesn't look like these links, e.g. `''animation trigger state/idle''` have linked to the definitions correctly.

ydaniv replied:
Interesting! I'm not sure whether this is unexpected. I guess we should open another issue for that.

Regarding the links, I'll check and fix.

ydaniv replied:
> Wouldn't setting did trigger to false this reset once triggers anytime the timeline goes idle (e.g. anytime a ScrollTimeline's scroller becomes large enough that it doesn't scroll)? I think this is probably unexpected.

Gave this another thought. I think this should be expected, but definitely worth opening an issue and discuss this explicitly.

> Note, it doesn't look like these links, e.g. ''animation trigger state/idle'' have linked to the definitions correctly.

Right, appears as though these aren't linked correctly, but I used the suggested syntax, so not sure what to do.

ydaniv replied:
The `did trigger` flag is used to differentiate between the initial idle state and the rest of the primary/inverse states going in/out of active interval. Mostly because you always want the initial state to be a non-triggered animation and the rest can be anything else defined by type: running/paused, played/reversed, etc.

I think it's OK for it to be reset on cases like timeline becoming inactive or if setting to a new trigger.
So not sure  whether the `once and only once` is actually called for.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2354,142 @@ Items sorted earlier are executed before those sorted later.
+
+### Animation Trigger Ranges ### {#trigger-ranges}
+
+An [=animation trigger=] has at least one [=animation attachment range|range=],
```
I think we should define a new term for the trigger interval's active range, as the combination of the two ranges would probably be worth eventually having a diagram to illustrate.

ydaniv replied:
Sounds good, I'll try adding that.

ydaniv replied:
done

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2354,142 @@ Items sorted earlier are executed before those sorted later.
+
+    The ''inverse'' behavior is [=reverse an animation|reversing=] the associated animation.
+
+  <dt><dfn>state</dfn>
```
I wonder if something like play-pause would be clearer.

ydaniv replied:
That's another option. The idea behind `state` was a reference to `play-state`.

flackr replied:
Right, I could see play-state as an alternative. State feels too strange, I didn't have any idea what this meant until I read the description :-).

flackr reviewed (changes requested):
Sorry for the delay on the review. Thanks for putting this together!

ydaniv commented on css-animations-2/Overview.bs:
```diff
@@ -690,7 +693,173 @@ The 'animation' shorthand property syntax is as follows:
+Canonical order: per grammar
+</pre>
+
+<span class=prod><dfn>&lt;single-animation-trigger-type></dfn> = once | repeat | alternate | state</span>
```
I was pondering all the `none`/`normal`/`auto` options, and I think we currently don't need them, since the initial value is `once`, with a timeline value of `auto`, which I think is exactly what we have today.
We could also add an issue on that, in case it needs an approval from the group.

ydaniv replied:
The definition of the values is done [here](https://github.com/w3c/csswg-drafts/pull/10128/files#diff-d4bbb6f14218df63c7727c884dba84e0103d39de0571ef1f7ab202c0a30e7608R2434).
Should I add here another description? More user/less implementor facing?

ydaniv replied:
> Can you add a description of what each property value means.

Done

ydaniv commented on css-animations-2/Overview.bs:
```diff
@@ -690,7 +693,173 @@ The 'animation' shorthand property syntax is as follows:
+Animation Type: not animatable
+</pre>
+
+Issue: Probably a trigger with timeline "none" is under-specified here. Need to clarify what it means.
```
Well, I remember we considered this in a previous issue specifically on syncing animations, but we said there that this will be the stuff that will eventually be done via GroupEffects. That said, if you see some way of changing the definition here and keeping this more open for later extensions I'm all for it.

ydaniv replied:
> but unlike animation delay could use absolute times since page load or other events I suppose?

I feel like we'll still the pseudo-selector you suggested in https://github.com/w3c/csswg-drafts/issues/8175, for example, to defer until DOM and assets are loaded. This is pretty common.

ydaniv replied:
OK, but `none` here is part of the `single-animation-timeline` syntax. Should we remove it just here? Or entirely?

flackr replied:
I see.. I suppose suppose we can keep it and I guess the animation would never start?

ydaniv replied:
Yeah, I guess it resets state back to `idle`.

ydaniv commented on web-animations-1/Overview.bs:
```diff
@@ -772,6 +772,9 @@ Animation Frames {#animation-frame-loop}
               [=associated with a timeline|associated with=] the timeline.
 
+          * Updating the [=animation trigger state=] of any [=animation trigger=]
+              [=associated with a timeline|associated with=] the timeline.
```
Well the `animation-frame-loop` definition exists only here, and I had to only add this one bullet. So what would be the proper way to add it to web-animations-2? Should I add just this delta somehow? Another way?

Regarding order, yes, I'll pull it up to happen before updating animations.

ydaniv replied:
done

ydaniv commented on css-animations-2/Overview.bs:
```diff
@@ -690,7 +693,173 @@ The 'animation' shorthand property syntax is as follows:
+    form a [=coordinating list property group=]
+    with 'animation-name' as the [=coordinating list base property=]
+    and each item in the [=coordinated value list=]
+    defining the properties of a single animation trigger.
```
I like both. I guess I'll refactor the text a bit and add a link.

ydaniv replied:
done

ydaniv commented:
@flackr kind ping, in case you missed this, I think it's ready for another round 🙏

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2376,152 @@ Items sorted earlier are executed before those sorted later.
+
+  <dt><dfn>state</dfn>
+  <dd>
+    The ''primary'' behavior is [=play an animation|triggering or resuming=] the associated animation.
```
I think it would be more consistent if state only changes the animation play state, i.e. the animation would be started in a paused state until the active trigger range is entered.

ydaniv replied:
Like I mentioned above, the idea was that when coming from `idle` state the animation is not triggered yet, and then upon entering `primary` state with `did trigger: false` the animation is triggered, and then only played/paused from that point.

ydaniv replied:
Looking again at `idle` state's definition, it's already stating how the animation starts:

> The [=animation effect=] associated |animation| remains in
    its [=animation effect/before phase=] and stays at zero [=animation/start time=].

So currently this is consistent across all types.
You think it will still be confusing for the `pause/play` behavior?

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2376,152 @@ Items sorted earlier are executed before those sorted later.
+
+  <dt><dfn>primary</dfn>
+  <dd>
+    The ''primary'' effect defined by |type| is applied to |animation|.
```
There's a bit of a mismatch between this and the text in css-animations-2 where the text for single-animation-trigger-type specifies when the effect is applied. The text above implies that the effect is applied continually while in the state, but the text in single-animation-trigger-type implies that the effect is applied at a discrete moment in time. If all of the effects are discrete then I think it would be simpler to have the effect explicitly only invoked when the trigger's state changes.

ydaniv replied:
I've tried making that more explicit in the text, LMK if that's good

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2376,152 @@ Items sorted earlier are executed before those sorted later.
+the [=animation trigger/active interval|active interval=] of an [=animation trigger=] |trigger| as follows:
+
+<dl class=switch>
+  : If |trigger| is inside its [=animation trigger/active interval|active interval=],
```
This should be based on the trigger state I think. Otherwise if you're in the exit range but not in the default range then this condition would be false.

ydaniv replied:
Right, refactored to use `|state|`

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2376,152 @@ Items sorted earlier are executed before those sorted later.
+[=Animation triggers=] define an <dfn for="animation trigger">active interval</dfn>,
+which is the period of time during which the trigger is scheduled to produce its effect
+on its associated [=animation=]’s playback.
+Each [=animation trigger=] has only one such interval,
```
This is super confusing, having *one* interval and then referring to two ranges. I think we should say that the active interval is determined by its most recent state.

ydaniv replied:
OK, I've rephrased the text to be more elaborate and explicit regarding dependency on state

ydaniv replied:
Actually, I'm refactoring all of it, and I'll add that as well.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2376,152 @@ Items sorted earlier are executed before those sorted later.
+<dl dfn-type=value dfn-for='animation trigger type'>
+  <dt><dfn>once</dfn>
+  <dd>
+    The ''primary'' behavior is [=play an animation|triggering=] the associated animation.
```
I think this needs to define how the behavior is only applied once. E.g. the suggested triggered state associated with an animation, initially false

ydaniv replied:
I think this is covered in the new state updating algo, right?

ydaniv replied:
I suppose if the type is `once` and `did trigger` is `true` it should suffice to indicate that the state should not change, right?

ydaniv commented on web-animations-2/Overview.bs:
```diff
@@ -221,6 +222,28 @@ Add:
+<h3 id="animation-frame-loop">Animation frames</h3>
+
+This section extends the definition of the [=update animations and send events=]
+procedure to include the following extra step:
```
Yeah, I think so. The intent is to add the additional task to this step as the first task.
How can I improve the text? Is there an example I can follow?

ydaniv replied:
I've removed that and instead added the algorithm to update the trigger's state as result of updating timeline's current time

flackr commented:
I think this also needs to add the animation trigger to [The Animation interface](https://drafts.csswg.org/web-animations-2/#the-animation-interface).

ydaniv commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2376,152 @@ Items sorted earlier are executed before those sorted later.
+
+The [=animation trigger state=] |state| of an [=animation trigger=] |trigger|
+is controlled by the [=animation attachment range|attachment range=] it is attached to,
+given a flag |did trigger|, as follows:
```
I've refactored the text below into an `algorithm` text

ydaniv commented:
@flackr I've added the definitions for the AnimationTrigger interface.
I've added a `Setting the timeline of an Animation Trigger` algorithm which should make the extra step in `update animations and send events` redundant and instead updating timelines' current time should take care of triggers as well.
I've also made some more fixes.
I think it's ready for another round, thanks!

ydaniv commented:
@flackr kind ping, should be ready for another review, thanks!

ydaniv commented:
@flackr any chance we could make some progress here to present something at TPAC?

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2355,191 @@ Items sorted earlier are executed before those sorted later.
+      : If |trigger|’s [=local time=] is [=unresolved=],
+      ::
+        1. Set |state| to ''animation trigger state/idle''.
+        1. Set |did trigger| to false.
```
I am still confused by this. We should not reset the state used to trigger `once` animations even if the local time is unresolved. We can get an unresolved time if the user changes the browser height for example such that the scroll timeline is no longer scrollable. This shouldn't result in re-arming a `once` animation.

ydaniv replied:
OK, I think it's fine to just remove this line. Then there's no way to reset it to `false`.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -3366,6 +3553,167 @@ Update the type for the AnimationPlaybackEventInit members.
+  attribute TimelineRangeOffset rangeStart;
+  attribute TimelineRangeOffset rangeEnd;
+  attribute TimelineRangeOffset exitRangeStart;
+  attribute TimelineRangeOffset exitRangeEnd;
```
I suspect these want to be the same complex type as rangeStart and rangeEnd: https://drafts.csswg.org/web-animations-2/#the-animation-interface

ydaniv replied:
done

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -3366,6 +3553,167 @@ Update the type for the AnimationPlaybackEventInit members.
+  TimelineRangeOffset? rangeStart = "normal";
+  TimelineRangeOffset? rangeEnd = "normal";
+  TimelineRangeOffset? exitRangeStart;
+  TimelineRangeOffset? exitRangeEnd;
```
I think rather than being nullable these all have default values just as their css counterparts do. Note that [TimelineRangeOffset](https://drafts.csswg.org/web-animations-2/#dictdef-timelinerangeoffset) is a structure and can't hold string values like "normal", but ([TimelineRangeOffset](https://drafts.csswg.org/web-animations-2/#dictdef-timelinerangeoffset) or [CSSNumericValue](https://drafts.css-houdini.org/css-typed-om-1/#cssnumericvalue) or [CSSKeywordValue](https://drafts.css-houdini.org/css-typed-om-1/#csskeywordvalue) or [DOMString](https://webidl.spec.whatwg.org/#idl-DOMString)) can.

ydaniv replied:
The thing about the exit range, is that you want it to by default inherit the default range values, and not reset to `normal`.
Maybe we should specify that with an `auto` value that will be the default?

ydaniv replied:
Oh, we already have that in the css counterparts, so adding it here as well.

ydaniv replied:
done

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -3366,6 +3553,167 @@ Update the type for the AnimationPlaybackEventInit members.
 
 </div>
 
+<h3 id="the-animationtrigger-interface">The <code>AnimationTrigger</code> interface</h3>
```
I think you still need to add AnimationTrigger to the animation interface: https://drafts.csswg.org/web-animations-2/#the-animation-interface as well as the KeyframeAnimationOptions used when constructing an animation: https://drafts.csswg.org/web-animations-1/#dictdef-keyframeanimationoptions

This would allow for an animation to be setup via the web-animations api doing something like this:
```js
element.animate({
    // slide and fade in
    transform: ['translateY(-50px)', 'none'],
    opacity: [0, 1]}, {
    duration: 1000,
    // Starts when the trigger point is reached.
    trigger: new AnimationTrigger(...)};
```

Note that trigger can be null, and if null you could apply a default trigger similar to how we use the default document timeline if timeline is null: https://drafts.csswg.org/web-animations-1/#dictdef-keyframeanimationoptions:~:text=if%20timeline%20member%20of%20options%20is%20missing

ydaniv replied:
In case it's null I think I rather keep it that way and not add a default trigger. Just keep the existing behavior. that should keep consistency with creating an Animation via constructor and inspecting the object.
WDYT?

ydaniv replied:
Done.
I've also added an explicit `setting a trigger of an animation` procedure that is called out wherever  a `trigger` is set in any related interface.

tabatkins commented on css-animations-2/Overview.bs:
```diff
@@ -690,7 +697,279 @@ The 'animation' shorthand property syntax is as follows:
+	<dt><dfn>repeat</dfn>
+	<dd>
+		The animation is played from the beginning each time it is triggered.
+		When the trigger’s [=animation trigger/active interval|active interval=] is exited the animation is reset.
```
```suggestion
		When the trigger’s [=animation trigger/active interval=] is exited the animation is reset.
```

tabatkins commented on css-animations-2/Overview.bs:
```diff
@@ -690,7 +697,279 @@ The 'animation' shorthand property syntax is as follows:
+	<dt><dfn>alternate</dfn>
+	<dd>
+		The animation is played forwards, according to its [=playback direction=], each time it is triggered.
+		When the trigger’s [=animation trigger/active interval|active interval=] is exited the animation is reversed.
```
```suggestion
		When the trigger’s [=animation trigger/active interval=] is exited the animation is reversed.
```

tabatkins commented on css-animations-2/Overview.bs:
```diff
@@ -690,7 +697,279 @@ The 'animation' shorthand property syntax is as follows:
+	<dt><dfn>state</dfn>
+	<dd>
+		The animation is triggered and played once.
+		When the trigger’s [=animation trigger/active interval|active interval=] is exited the animation is paused.
+		When the trigger’s [=animation trigger/active interval|active interval=] is re-entered the animation is resumed.
```
```suggestion
		When the trigger’s [=animation trigger/active interval=] is exited the animation is paused.
		When the trigger’s [=animation trigger/active interval=] is re-entered the animation is resumed.
```

tabatkins reviewed (commented):
(just some super minor markup nits with suggestions)

ydaniv commented:
@flackr I think all last comments are addressed, thanks!

ydaniv commented:
@flackr there was a missing definition for the `Animation/trigger` attribute. Added it now.

ydaniv commented:
@flackr kind ping, I think this is ready for review.

ydaniv commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2355,190 @@ Items sorted earlier are executed before those sorted later.
+### Updating Animation Trigger State ### {#updating-trigger-state}
+
+<div algorithm="updating animation trigger state">
+  The procedure of <dfn>updating animation trigger state</dfn> |state|
```
Well the idea was that it follows naturally from updating that timelines, but certainly we could note that explicitly on the note under 1.

flackr replied:
Yes, that'd be great to amend the note to explicitly call this out.

ydaniv replied:
Done

ydaniv commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2355,190 @@ Items sorted earlier are executed before those sorted later.
+
+  1. Set |state| as follows:
+    <dl class=switch>
+      : If |trigger|’s [=local time=] is [=unresolved=],
+      ::
+        1. Set |state| to ''animation trigger state/idle''.
```
@flackr I think this is also problematic, since the timeline may switch between active/inactive and `state` may switch between `idle` and `primary` accordingly, since we don't check for `did trigger` here.
According to what we agreed we don't want the `state` to go back to `idle` after `did trigger=true`, right? So perhaps it would be better to say that `state` is initially `idle`, and then not set it on that step above again, only abort the procedure.
WDYT?

flackr replied:
Yes, I remember talking about this before. I agree the state should be initially idle on creation but not go back to being idle as a result of an update.

ydaniv replied:
Done

ydaniv commented:
@flackr just pushed fixes for the last round of reviews, I think it's ready for another check.

DavMila commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2362,204 @@ Items sorted earlier are executed before those sorted later.
+      ::
+        The ''primary'' behavior is [=play an animation|triggering=] the associated animation.
+
+      : Otherwise,
+      ::
+        The ''primary'' behavior is [=reverse an animation|reversing=] the associated animation.
```
A little curious about how this works. Am I correct that for `animation-trigger-type: alternate` the animation should be played when entering the `animation-trigger-range` and reversed when exiting the `animation-trigger-exit-range`? If so, it seems like the behavior defined for `alternate` here could result in reversing the animation while we're inside the exit range?

Let me try to use an example and perhaps you can point out what I've misunderstood. Say we have:
```
ainmation-trigger: view() alternate cover 0% cover 100%;
```
and lets say cover 0% and cover 100% correspond to 100px and 200px respectively.
So the default range and the exit range are both [100, 200].
The page loads, `scrollTop`= 0, the trigger's state is `idle` (active interval is [100,200]).
Then we scroll to 150px, so we set the trigger's state to `primary` (the active interval is now the exit range, still [100, 200]. since `did trigger` is false we play the animation) and set `did trigger` to true.
We then scroll to 160px and the animation trigger's state is updated: since we are still inside the active interval,
we set the state to `primary` (*since `did trigger` is true, we reverse the animation*) and set `did trigger` to true.
So we would have reversed the animation despite not having exited the exit range?
Or perhaps I'm not quite correct about when the animation trigger gets updated?
It seems like we don't need the if/else and the `alternate` behavior should just be `primary`->`trigger`, `inverse`->`reverse`?
I might also have missed something about `did trigger`... it seems the algorithm is given a flag `did trigger`, but also defines a `did trigger` var, and the animation trigger also has an internal `did trigger`, so maybe I should clarify how those are related.

ydaniv replied:
Yeah, this one is a bit tricky. The idea is that if `did trigger` is false then we play "forwards". Then, on every state change we reverse. The problem,  as you described it, is that state change will occur on every scroll change. But my internet wad that a state change occurs _only_ when the active attachment range is either centered or exited.
So the way I intend for it behave is to not change state until the range is exited. Then reverse on exit, and reverse back on re-enter, and so on 

Does this make sense?

DavMila replied:
I might not be totally clear on the last line:
>  Then reverse on exit, and reverse back on re-enter, and so on

I understand that we would reverse the animation on exit (exiting the exit range), but on re-enter (re-entering the trigger range) we would just play the animation forwards again right? is this what you mean by "reverse back"?

If that's right, what do you think about simplifying the `primary` behavior for `alternate`? I think that simply having:
```
'alternate'
   The primary behavior is triggering the associated animation. The inverse behavior is reversing the associated animation.
```
fixes the issue of reversing on every scroll?

ydaniv replied:
> I understand that we would reverse the animation on exit (exiting the exit range), but on re-enter (re-entering the trigger range) we would just play the animation forwards again right? is this what you mean by "reverse back"?

What I mean is that after reversing an animation its effective playback rate remains inveresed, so you'll first need to inverse the playback rate, and then invoke the play an animation procedure, which is exactly the [reversing an animation procedure](https://www.w3.org/TR/web-animations-1/#reversing-an-animation-section).

> fixes the issue of reversing on every scroll?

My issue with "every scroll" is that I'd just like to clarify that we don't do anything on every scroll change. Even though we're using Scroll and View timelines here, the changes are singular on range entry/exit, and don't occur repeatedly on progress change, like SDA's do. Otherwise we would need to re-trigger `primary` on each scroll change.

flackr replied:
Yes, the "reverse" on re-enter is the forwards direction, however, it is worth noting that it might not start from the beginning of the animation if the reverse animation hadn't finished. E.g. if you temporarily move out of the exit range and then back into the trigger range you may only play from backwards from 100% back to 80%, then reverse back to 100%.

ydaniv replied:
Yes, of course. This is the precise intent, just like flipping a transition mid-way.

DavMila commented on web-animations-2/Overview.bs:
```diff
@@ -2353,6 +2362,204 @@ Items sorted earlier are executed before those sorted later.
+                  1. Set |state| to ''animation trigger state/primary''.
+                  1. Set |did trigger| to true.
+
+              : Otherwise,
```
In addition to my comment above, would we have an issue if we have non-overlapping trigger and exit ranges? One use case I can think of for non-overlapping ranges is if you wanted a `repeat` trigger but in only one scrolling direction. E.g. you only want the animation to trigger scrolling downwards (item translates upwards into view). In this case you'd define the exit range and the default range as, e.g. [0, 100] and [150, infinity] so you can only reset the animation by scrolling back up by a sufficient amount. As currently written it looks like this algorithm will probably reset the animation shortly after we trigger it:
- enter the trigger range, e.g. 160px
- - plays the animation, state becomes `primary` -> exit range [0, 100] becomes active interval
- further scrolling (still outside exit range, e.g. 170px)
- - animation is reset.
?

One way we could address this would be to define two types of (active) intervals, an `entry interval` and an `exit interval` and have the trigger keep track of whether, at the last update it was within the interval `within active interval` defaulting to false. Then the algorithm can decide whether to apply the `primary` or `inverse` behavior based on whether we are entering or exiting the relevant interval.

ydaniv replied:
I'm a little confused here.
Let's first see we're on the same page about the exit range:
Once the user is inside the default range, and `primary` behavior is triggered, the active range is immediately switched to the exit range. And only once you exit that range, the state switches again.
If we allow exit ranges that are "smaller" then their corresponding default ranges then we could be allowing endless loops or other weird behaviors. So I suggested as an issue in the PR to perhaps restrict exit ranges to *only* be equal or larger than default ranges, and that they can not constrict them, see [here](https://github.com/w3c/csswg-drafts/pull/10128/files#diff-d4bbb6f14218df63c7727c884dba84e0103d39de0571ef1f7ab202c0a30e7608R2461).

I can't think now about any interesting effect one could do with a non-overlapping exit range, so I suggest avoiding it, just like we enforce range start is not larger than range end, or the timeline becomes inactive.

DavMila replied:
Okay, thanks for adding that as an issue in the spec. I guess it (restricting  valid values of the exit range to be "equal to or greater than" the trigger range) is something the working group will want to consider and we can file an issue for it after this PR has landed.

ydaniv replied:
Ok, they both define the active range. I'm not following. Could you be more specific technically what would that mean in the spec?

DavMila replied:
> I think to make things simple we may just want to spec that being in the trigger range also counts as being in the exit range.

Hmm, could you elaborate? I'm not sure that will fully address the concern about how the current PR will handle non-overlapping ranges, particularly if I changed the entry range from [150, infinity] to [150, 200] and scrolling beyond 200px is possible.

DavMila replied:
> > I think to make things simple we may just want to spec that being in the trigger range also counts as being in the exit range.
> 
> Hmm, could you elaborate? I'm not sure that will fully address the concern about how the current PR will handle non-overlapping ranges, particularly if I changed the entry range from [150, infinity] to [150, 200] and scrolling beyond 200px is possible (ignoring that that now makes it possible to trigger the animation by scrolling either up or down).

Ah.. but maybe it's not such a valid use case anymore if you shrink the entry range - you can trigger it scrolling up or down, but you can only reset it scrolling up far enough and then scrolling down again :)

flackr replied:
@ydaniv I haven't re-reviewed the specific algorithm, but the idea is we wouldn't enter the state/inverse state if you're in the primary trigger range - which means it doesn't matter if your exit range is defined to be inclusive of it or not, the entry range is by definition part of it.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2488,15 +2488,18 @@ Depending on its most recent [=animation trigger state|state=] |state|, the
     <dl class=switch>
       : If |trigger|’s [=local time=] is [=unresolved=],
       ::
-        1. Set |state| to ''animation trigger state/idle''.
```
Would it make sense to count this as not being inside the trigger range, i.e. switching to state/inverse if we've triggered? It might be a bit odd that you stay in the triggered state when the timeline is inactive.

flackr reviewed (approved):
I think this is looking good. There are still a few details to work out but overall I think the general algorithm should work pretty well and I don't want to delay getting the initial spec in :-)
